binary_transparency/firmware/cmd/ft_monitor/impl/ft_monitor.go

```diff
@@ -89,60 +103,77 @@ func Main(ctx context.Context, opts MonitorOpts) error {
+	stmt := entry.Value
+	if stmt.Type != api.FirmwareMetadataType {
+		// Only analyze firmware statements in the monitor.
+		return
+	}
 
-		// Fetch the Image from FT Personality
-		image, err := c.GetFirmwareImage(meta.FirmwareImageSHA512)
-		if err != nil {
-			glog.Warningf("Unable to GetFirmwareImage for Firmware with Hash 0x%x , reason %q", meta.FirmwareImageSHA512, err)
-			continue
-		}
-		// Verify Image Hash from log Manifest matches the actual image hash
-		h := sha512.Sum512(image)
-		if !bytes.Equal(h[:], meta.FirmwareImageSHA512) {
-			glog.Warningf("downloaded image does not match SHA512 in metadata (%x != %x)", h[:], meta.FirmwareImageSHA512)
-			continue
-		}
-		glog.V(1).Infof("Image Hash Verified for image at leaf index %d", entry.Index)
+	// Parse the firmware metadata:
+	var meta api.FirmwareMetadata
+	if err := json.Unmarshal(stmt.Statement, &meta); err != nil {
+		glog.Warningf("Unable to decode FW Metadata from Statement %q", err)
+		return
+	}
 
-		// Search for specific keywords inside firmware image
-		malwareDetected := matcher.Match(image)
+	glog.Infof("Found firmware (@%d): %s", entry.Index, meta)
 
-		if malwareDetected {
-			opts.Matched(entry.Index, meta)
+	// Fetch the Image from FT Personality
+	image, err := c.GetFirmwareImage(meta.FirmwareImageSHA512)
+	if err != nil {
+		glog.Warningf("Unable to GetFirmwareImage for Firmware with Hash 0x%x , reason %q", meta.FirmwareImageSHA512, err)
+		return
+	}
+	// Verify Image Hash from log Manifest matches the actual image hash
+	h := sha512.Sum512(image)
+	if !bytes.Equal(h[:], meta.FirmwareImageSHA512) {
+		glog.Warningf("downloaded image does not match SHA512 in metadata (%x != %x)", h[:], meta.FirmwareImageSHA512)
+		return
+	}
+	glog.V(1).Infof("Image Hash Verified for image at leaf index %d", entry.Index)
+
+	// Search for specific keywords inside firmware image
+	malwareDetected := matcher.Match(image)
+
+	if malwareDetected {
+		opts.Matched(entry.Index, meta)
+	}
+	if opts.Annotate {
+		ms := api.MalwareStatement{
+			FirmwareID: api.FirmwareID{
+				LogIndex:            entry.Index,
+				FirmwareImageSHA512: h[:],
+			},
+			Good: !malwareDetected,
 		}
-		if opts.Annotate {
-			ms := api.MalwareStatement{
-				FirmwareID: api.FirmwareID{
-					LogIndex:            entry.Index,
-					FirmwareImageSHA512: h[:],
-				},
-				Good: !malwareDetected,
-			}
-			glog.V(1).Infof("Annotating %s", ms)
-			js, err := createStatementJSON(ms)
-			if err != nil {
-				return fmt.Errorf("failed to create annotation: %w", err)
-			}
-			sc := client.SubmitClient{
-				ReadonlyClient: &c,
-			}
-			if err := sc.PublishAnnotationMalware(js); err != nil {
-				return fmt.Errorf("failed to publish annotation: %w", err)
-			}
+		glog.V(1).Infof("Annotating %s", ms)
+		js, err := createStatementJSON(ms)
+		if err != nil {
+			glog.Warningf("failed to create annotation: %q", err)
```

Comment:
The earlier Error points in the old code on these have been turned now into Warnings.
Should we not fail, if the statement creation fails?

Reply:
Good spot. I've made these checks consistent with the other checks in the monitor, which were all warnings. The history of this class is that initially everything was a warning, and then I inadvertently introduced the dissonance when I added the annotation code and made things throw errors. The refactoring in this PR made that dissonance clear, so I've resolved it by making the annotation code consistent with everything else.

If we want to, we could follow up with a new PR to make all of these warnings be errors. I'm not sure about that though. For demoware, this strikes a reasonable middle ground.

Reply:
I will be OK, if we track it with a new issue, to classify which of these warnings are genuine errors which we should block monitor from further operation, versus which we can continue to proceed as Warnings. This way we can revisit this section of code and then proceed with this PR in the state as it is, marking the comment to resolved!

Reply:
I've changed the code to return errors from all places, and the calling code then logs the warning. There is a TODO there to consider a strict processing mode which will exit instead of logging the warning. I think this is a reasonable tidy compromise?

Reply:
Thanks, Yes much better now!